File: android/app/src/main/kotlin/au/radar/app/AlertVoice.kt
package au.radar.app

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.ToneGenerator
import android.speech.tts.TextToSpeech
import au.radar.core.Announcement
import au.radar.core.AnnouncementLevel
import java.util.Locale

/**
 * Speaks warnings over whatever the driver is listening to.
 *
 * The audio attributes are the fiddly part: warnings must duck music rather
 * than stop it, and must route to the car over Bluetooth and Android Auto.
 * USAGE_ASSISTANCE_NAVIGATION_GUIDANCE is what tells the system this is a
 * navigation prompt and should be treated like one.
 */
class AlertVoice(context: Context) {

    private var ready = false
    private val tone = ToneGenerator(AudioManager.STREAM_NOTIFICATION, 80)

    private val engine = TextToSpeech(context.applicationContext) { status ->
        ready = status == TextToSpeech.SUCCESS
        if (ready) {
            // Australian English first so road names and "kilometres" land right.
            val result = tts.setLanguage(Locale("en", "AU"))
            if (result == TextToSpeech.LANG_MISSING_DATA || result == TextToSpeech.LANG_NOT_SUPPORTED) {
                tts.language = Locale.UK
            }
            // Slightly faster than default: a warning that takes three seconds
            // to deliver at 100 km/h has already cost eighty metres.
            tts.setSpeechRate(1.05f)
        }
    }

    private val tts: TextToSpeech get() = engine

    init {
        engine.setAudioAttributes(
            AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_ASSISTANCE_NAVIGATION_GUIDANCE)
                .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                .build(),
        )
    }

    fun announce(announcement: Announcement) {
        when (announcement.level) {
            AnnouncementLevel.CHIME -> tone.startTone(ToneGenerator.TONE_PROP_BEEP, 180)
            AnnouncementLevel.SPEAK -> speak(announcement.spokenText)
        }
    }

    private fun speak(text: String) {
        if (!ready) return
        // QUEUE_FLUSH, not QUEUE_ADD: a backlog of warnings is worse than none,
        // because the one being read is no longer the one you are approaching.
        tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, text.hashCode().toString())
    }

    /**
     * Turn-by-turn instructions, which bypass the alert engine's suppression:
     * a maneuver is timed to the road, not rationed like a hazard warning.
     */
    fun speakNavigation(text: String) {
        speak(text)
    }

    fun release() {
        tts.stop()
        tts.shutdown()
        tone.release()
    }
}

File: ios/App/Sources/AlertVoice.swift
import AVFoundation
import Foundation
import RadarKit

/// Speaks warnings over whatever the driver is listening to.
///
/// The audio session is the fiddly part: the app has to duck music rather than
/// stop it, has to work over CarPlay and Bluetooth, and must not steal the
/// session permanently when it has nothing to say.
final class AlertVoice {
    private let synthesizer = AVSpeechSynthesizer()
    private var sessionActive = false

    init() {
        configureSession()
    }

    private func configureSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(
                .playback,
                mode: .voicePrompt,
                options: [.duckOthers, .interruptSpokenAudioAndMixWithOthers, .allowBluetooth]
            )
        } catch {
            // Without a session we simply stay silent rather than crash mid-drive.
        }
    }

    func announce(_ announcement: Announcement) {
        switch announcement.level {
        case .chime:
            playChime()
        case .speak:
            speak(announcement.spokenText)
        }
    }

    private func speak(_ text: String) {
        activateSession()
        let utterance = AVSpeechUtterance(string: text)
        utterance.voice = AVSpeechSynthesisVoice(language: "en-AU")
            ?? AVSpeechSynthesisVoice(language: "en-GB")
        // Slightly faster than default: a warning that takes three seconds to
        // deliver at 100 km/h has cost you eighty metres.
        utterance.rate = AVSpeechUtteranceDefaultSpeechRate * 1.05
        utterance.postUtteranceDelay = 0.1
        synthesizer.speak(utterance)
    }

    private func playChime() {
        activateSession()
        // 1057 is the short system "tock". A distinct tone file can replace this
        // later; the point is that a chime is not a sentence.
        AudioServicesPlaySystemSound(1057)
    }

    private func activateSession() {
        guard !sessionActive else { return }
        try? AVAudioSession.sharedInstance().setActive(true)
        sessionActive = true
    }

    /// Hand the audio session back when the drive ends, so music un-ducks.
    func release() {
        guard sessionActive else { return }
        try? AVAudioSession.sharedInstance().setActive(
            false, options: .notifyOthersOnDeactivation
        )
        sessionActive = false
    }
}
